### PR TITLE
Crawlers

### PR DESCRIPTION
**Pull request by wdm0006**
Type: Enhancement

## Changes

- Moved webpage into its own file
- Moved engines into a subdirectory
- Added a subdirectory for crawlers
- Added a default crawler that just loops through a list of URLs
- Added an examples directory with an example using the crawler
- Updated docs.

## Documentation

The intention of this PR is to set the stage for full fledged crawler development, so general project structuring, and the definition of an interface UX for interacting with crawlers and engines. 

## Impacted Areas

- Pretty much everything.

## Checklist

- [x] Add new feature to the usage documentation
- [x] Add tests

### CODE REVIEW

File: rasp/crawlers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+from rasp.crawlers.base import DefaultBoundedCrawler
+
+__all__ = [
+    "DefaultBoundedCrawler"
+]
```
Comment (dlrobertson) [inline rasp/crawlers/__init__.py]: nit: newline

**Review by dlrobertson — APPROVED** [review]
Nice! The more I think about it the more I like the Bounded and UnBounded idea.

File: rasp/engines/base.py [hunk 2]
```diff
@@ -67,7 +66,7 @@ def as_func(cls, method_name):
         try:
             return getattr(TmpEngine, method_name)
         except AttributeError:
-            raise NotImplementedError('Class {} does not implement {}'.format(class_name, func_name))
+            raise NotImplementedError('Class {} does not implement {}'.format(class_name, method_name))
```
Comment (hank-bond) [inline rasp/engines/base.py]: Already fixed this [here](https://github.com/anidata/rasp/pull/26/files#diff-3387bc458c72e04ddfef789b19a5bcafR71).  Make sure to rebase/merge your branch on the current develop.

File: rasp/crawlers/base.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class BaseBoundedCrawler(object):
+    def run(self):
+        raise NotImplemented("run not implemented for {}"
+                             .format(str(self.__class__.__name__)))
+
+
+class DefaultBoundedCrawler(BaseBoundedCrawler):
```
Comment (hank-bond) [inline rasp/crawlers/base.py:7]: Can you write some quick docs on the crawler?  Also RT what would an UnBounded crawler look like?  Is that just a stateless crawler?

File: examples/basic_crawler.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from rasp import DefaultEngine, DefaultBoundedCrawler
+import time
+
+
+def example_callback(wp):
+    print(wp.url)
+
+if __name__ == '__main__':
```
Comment (hank-bond) [inline examples/basic_crawler.py:8]: Can we also add tests to the suite for the crawler?

**Review by hank-bond — COMMENTED** [review]
I'll re review when you have the conflicts fixed.  I like the direction of this PR, good reorganization.

File: rasp/crawlers/base.py (hunk 3 above)
Comment (wdm0006) [reply rasp/crawlers/base.py:7]: Unbounded in my mind would sit on an external queue, which some thing was putting messages into and the crawler would be able to submit new urls to that queue as well.  So a bit more involved.  I'll pull in develop and make other changes when I get a chance, thanks for the notes.

**Comment by wdm0006** [comment]
Fixed merge conflicts, added a little bit to the docs and readme, fixed the existing tests that I broke, and wrote a small new one.

**Comment by wdm0006** [comment]
Going to go ahead and close this, since all of @baronvonbadguy's points were addressed.

**Pull request merged**